ACE/ace/RAW_Socket.cpp

```diff
@@ -0,0 +1,414 @@
+                                      n,
+                                      flags,
+                                      (sockaddr *) saddr,
+                                      &addr_len);
+
+    addr.set_size (addr_len);
+    addr.set_type (saddr->sa_family);
+    return status;
+}
+
+static inline void fillMsgHdr(msghdr& recv_msg, const ACE_INET_Addr &addr, void* pcbuf, int cbuf_size)
+{
+  #if defined (ACE_HAS_SOCKADDR_MSG_NAME)
+    recv_msg.msg_name = static_cast<struct sockaddr *> (addr.get_addr ());
+  #else
+    recv_msg.msg_name = static_cast<char *>(addr.get_addr ());
+  #endif /* ACE_HAS_SOCKADDR_MSG_NAME */
+  
+  recv_msg.msg_namelen = addr.get_size ();
+
+  #ifdef ACE_USE_MSG_CONTROL
+    recv_msg.msg_control    =  pcbuf;
+    recv_msg.msg_controllen =  cbuf_size;
+  #elif !defined ACE_LACKS_SENDMSG
+    recv_msg.msg_accrights    = 0;
+    recv_msg.msg_accrightslen = 0;
+  #endif
+
+}
+
+static inline void getToAddrFromMsgHdr(msghdr& recv_msg, ACE_INET_Addr& to_addr)
+{
+  #ifdef ACE_USE_MSG_CONTROL
+        if (to_addr.get_type() == AF_INET) {
+    #if defined (IP_RECVDSTADDR) || defined (IP_PKTINFO)
+          for (cmsghdr *ptr = ACE_CMSG_FIRSTHDR (&recv_msg); ptr; ptr = ACE_CMSG_NXTHDR (&recv_msg, ptr)) {
+            #if defined (IP_RECVDSTADDR)
+            if (ptr->cmsg_level == IPPROTO_IP && ptr->cmsg_type == IP_RECVDSTADDR) {
+              to_addr.set_address (reinterpret_cast<char *>(ACE_CMSG_DATA (ptr)),
+                                    sizeof (struct in_addr),
+                                    0);
+              break;
+            }
+            #else
+            if (ptr->cmsg_level == IPPROTO_IP && ptr->cmsg_type == IP_PKTINFO) {
+              to_addr.set_address ((const char *) &(((in_pktinfo *) (ACE_CMSG_DATA (ptr)))->ipi_addr),
+                                    sizeof (struct in_addr),
+                                    0);
+              break;
+            }
+            #endif
+          }
+    #endif
+        }
+
+    #if defined (ACE_HAS_IPV6) && defined (IPV6_PKTINFO)
+        else if (to_addr.get_type() == AF_INET6) {
+          for (cmsghdr *ptr = ACE_CMSG_FIRSTHDR (&recv_msg); ptr; ptr = ACE_CMSG_NXTHDR (&recv_msg, ptr)) {
+            if (ptr->cmsg_level == IPPROTO_IPV6 && ptr->cmsg_type == IPV6_PKTINFO) {
+              to_addr.set_address (reinterpret_cast<char *>(&((reinterpret_cast<in6_pktinfo*>((ACE_CMSG_DATA(ptr))))->ipi6_addr)),
+                                    sizeof (struct in6_addr),
+                                    0);
+
+              break;
+            }
+          }
+        }
+    #endif
+  #endif
+
+}
+
+ssize_t
+ACE_RAW_SOCKET::recv (void *buf,
+                      size_t n,
+                      ACE_INET_Addr &addr,
+                      int flags,
+                      const ACE_Time_Value *timeout,
+                      ACE_INET_Addr *to_addr) const
+{
+  ACE_TRACE ("ACE_RAW_SOCKET::recv");
+
+  RECV_EXCEPTION_RETURN();
```

Comment:
I don't like the macros for this, makes it very hard to see what is checked

Reply:
The checking conditions are unlikely and trivial, so  need  not care what is the exceptions . 
The common macro trick can return the flow in place without any further if statement and can reduce the cyclomatic complexity